docs/proposals/synchronous-garbage-collection.md

```diff
@@ -0,0 +1,203 @@
+     BlockOwnerDeletion *bool
+}
+```
+
+The initial draft of the proposal did not include this field and it had a security loophole: a user who is only authorized to update one resource can set ownerReference to block the synchronous GC of other resources. Requiring users to explicitly set `BlockOwnerDeletion` allows the master to properly authorize the request.
+
+## DeleteOptions
+
+```go
+DeleteOptions {
+  …
+  // Whether and how garbage collection will be performed.
+  // Defaults to DefaultPropagationPolicy
+  PropagationPolicy *DeletePropagationPolicy
+}
+
+type DeletePropagationPolicy string
+
+const (
+    // Respects the existing garbage collection related finalizers on the object and the default garbage collection policy of the resource.
+    DefaultPropagationPolicy DeletePropagationPolicy = "Default"
+    // Orphans the dependents
+    OrphanDependents DeletePropagationPolicy = "Orphan"
+    // Deletes the object from the key-value store, the garbage collector will delete the dependents in the background.
+    DeleteDependentsInBackground DeletePropagationPolicy = "DeleteDependentsInBackground"
+    // The object exists in the key-value store until the garbage collector deletes all the dependents whose ownerReference.blockOwnerDeletion=true from the key-value store.
+    // API sever will put the "DeletingDependents" finalizer on the object, and sets its deletionTimestamp.
+    // This policy is cascading, i.e., the dependents will be deleted with GarbageCollectionSynchronous.
+    DeleteAfterBlockingDependentsAreDeleted DeletePropagationPolicy = "DeleteAfterBlockingDependentsAreDeleted"
+)
+```
+
+The `DeleteAfterBlockingDependentsAreDeleted` policy represents the synchronous GC mode.
+
+`DeleteOptions.OrphanDependents *bool` will be marked as deprecated and will be removed in 1.7. Validation code will make sure only one of `OrphanDependents` and `PropagationPolicy` may be set. We decided not to add another `DeleteAfterDependentsDeleted *bool`, because together with `OrphanDependents`, it will result in 9 possible combinations and is thus confusing.
+
+The conversion rules are described in the following table:
+
+| 1.5                                      | pre 1.4/1.4              |
+|------------------------------------------|--------------------------|
+| DefaultPropagationPolicy                 | OrphanDependents==nil    |
+| OrphanDependents                         | *OrphanDependents==true  |
+| DeleteDependentsInBackground             | *OrphanDependents==false |
+| DeleteAfterBlockingDependentsAreDeleted  | N/A                      |
+
+# Components changes
+
+## API Server
+
+`Delete()` function checks `DeleteOptions.PropagationPolicy`. If the policy is `DeleteAfterBlockingDependentsAreDeleted`, the API server will update the object instead of deleting it, add the finalizer, and set the `ObjectMeta.DeletionTimestamp`.
```

Comment:
(i.e., if it's present and they ask for orphan or background gc, we need to remove it)